design/anova/anova.Rmd

```diff
@@ -0,0 +1,395 @@
+2. If necessary, equate new symbols to compound expressions in the F1 block
+3. Equate all symbolic coefficients outside F1 block to a linear function of the symbols in the F1 block.
+
+In practice, if a effect is not contained in any other tests, use its orthogonal form (identity matrix) in the corresponding submatrix, e.g. for `AVISIT * PARMAMCD`
+If a effect is contained in other effects, for that part we try to equate the values.
+
+In our case of design matrix, testing `ARMCD`, there are 4 levels in total for `AVISIT * ARMCD`.
+The coefficients for the effect `ARMCD` is 1, so each value for the `AVISIT * ARMCD` is 1/4.
+
+For each level of `AVISIT`, there are 2 levels for `AVISIT * ARMCD`.
+So the coefficients for `AVISIT * ARMCD` is 1/2.
+
+However, this only works for the simplest case that there is no restriction on the design matrix.
+In general, we will not have such cases so it is fine.
+
+Note: using Type II on perfectly balanced data, the result should be the same for Type III.
+
+### ARMCD part
+```{r}
+l <- rep(0, 8) # initial values
+l[2] <- 1
+l[6:8] <- 1 / 4
+
+df_md(fit, l) # very close !
+```
+
+### AVISIT part
+
+```{r}
+l <- matrix(0, nrow = 3, ncol = 8) # initial values
+l[, c(3, 4, 5)] <- diag(rep(1, 3))
+l[, c(6, 7, 8)] <- diag(rep(1, 3)) / 2
+
+df_md(fit, l) # also close
+```
+
+
+# Prototypes
+
+```{r}
+get_l_matrix_type2 <- function(fit, var) {
+  mx <- component(fit, "x_matrix")
+  asg <- attr(mx, "assign")
+  formula <- fit$formula_parts$model_formula
+  tms <- terms(formula)
+  fcts <- attr(tms, "factors")[-1, , drop = FALSE] # discard the response
+  ods <- attr(tms, "order")
+  idx <- which(var == colnames(fcts))
+  cols <- which(asg == idx)
+  all_vars <- row.names(fcts)[which(fcts[, idx] == 1)]
+  all_levels <- vapply(all_vars, function(x) {
+    length(attr(fit$tmb_data$data[[x]], "levels"))
+  }, FUN.VALUE = 1)
+  total_levels <- prod(all_levels)
+  lvls <- prod(all_levels - 1)
+  l_mx <- matrix(0, nrow = lvls, ncol = ncol(mx))
+  l_mx[, cols] <- diag(rep(1, length(cols)))
+  for (i in seq_len(ncol(fcts))) {
+    x1 <- mx[, cols, drop = FALSE]
+    if (ods[i] >= ods[idx] && all(fcts[, i] >= fcts[, idx])) {
+      x0 <- mx[, which(!asg %in% c(i, idx)), drop = FALSE]
+      x2 <- mx[, which(asg == i), drop = FALSE]
+      m <- diag(rep(1, nrow(x0))) - x0 %*% solve(t(x0) %*% x0) %*% t(x0)
+      l_coefs <- solve(t(x1) %*% m %*% x1) %*% t(x1) %*% m %*% x2
+      l_mx[, which(asg == i)] <- l_coefs
+    }
+  }
+  l_mx
+}
+
+get_l_matrix_type3 <- function(fit, var) {
+  mx <- component(fit, "x_matrix")
+  asg <- attr(mx, "assign")
+  formula <- fit$formula_parts$model_formula
+  tms <- terms(formula)
+  fcts <- attr(tms, "factors")[-1, , drop = FALSE] # discard the response
+  ods <- attr(tms, "order")
+  idx <- which(var == colnames(fcts))
+  cols <- which(asg == idx)
+  all_vars <- row.names(fcts)[which(fcts[, idx] == 1)]
```

Comment:
quite some duplicated code between the two functions - can we split this out in a helper function that is used by both?

Reply:
yes this is only a prototype that works. not necessarily be the final function.

Reply:
yeah sure, just it might be helpful to already do the split up now in the design

Reply:
it is in the same function now